luarules/gadgets/unit_smartfight.lua

```diff
@@ -0,0 +1,406 @@
+	local spIsUnitInLos 		= Spring.IsUnitInLos
+	
+	local gaiaTeamID = spGetGaiaTeamID()
+	
+	local commandUpdateQueue = {}
+	local activeUnits = {}
+	
+	local activeCapturers = {}
+	local protectedTargets = {}
+
+	--------------------------------------------------------------------------------
+	-- INITIALIZATION
+	--------------------------------------------------------------------------------
+	function gadget:Initialize()
+		gadgetHandler:RegisterCMDID(CMD_SMARTFIGHT)
+		
+		-- Watch weapons for mobile units only (Optimization)
+		for unitDefID, unitDef in pairs(UnitDefs) do
+			if unitDef.canMove and (not unitDef.isBuilding) and unitDef.weapons then
+				for _, weapon in ipairs(unitDef.weapons) do
+					if weapon.weaponDef then
+						Script.SetWatchWeapon(weapon.weaponDef, true)
+					end
+				end
+			end
+		end
+	end
+
+	--------------------------------------------------------------------------------
+	-- GLOBAL SAFETY
+	--------------------------------------------------------------------------------
+	function gadget:UnitCommand(unitID, unitDefID, teamID, cmdID, cmdParams, cmdOptions)
+		if cmdID == CMD_CAPTURE then activeCapturers[unitID] = teamID end
+	end
+
+	function gadget:UnitCmdDone(unitID, unitDefID, teamID, cmdID, cmdTag, cmdParams, cmdOptions)
+		if cmdID == CMD_CAPTURE then activeCapturers[unitID] = nil end
+	end
+
+	local function UpdateProtectedTargets()
+		protectedTargets = {}
+		for builderID, teamID in pairs(activeCapturers) do
+			local isCapturing = false
+			-- Ensure builder still exists and is still capturing
+			if spValidUnitID(builderID) then
+				local cmd = spGetUnitCommands(builderID, 1)
+				if cmd and cmd[1].id == CMD_CAPTURE then
+					local targetID = cmd[1].params[1]
+					if targetID then
+						protectedTargets[targetID] = teamID
+						isCapturing = true
+					end
+				end
+			end
+			if not isCapturing then activeCapturers[builderID] = nil end
+		end
+	end
+
+	function gadget:AllowWeaponTarget(attackerID, targetID, attackerWeaponNum, attackerWeaponDefID, defaultPriority)
+		-- 1. Safety Checks (Nil Handling)
+		if not targetID or not attackerID then return true, defaultPriority end
+
+		-- 2. Fast Lookup: Is this target protected?
+		local capturingTeam = protectedTargets[targetID]
+		if not capturingTeam then return true, defaultPriority end
+
+		-- 3. Anti-Grief: Only block if Attacker & Capturer are SAME TEAM
+		local attackerTeam = spGetUnitTeam(attackerID)
+		if not attackerTeam then return true, defaultPriority end
+		
+		if attackerTeam ~= capturingTeam then return true, defaultPriority end
+		
+		-- 5. Block Fire
+		return false
+	end
+
+	--------------------------------------------------------------------------------
+	-- SMART FIGHT LOGIC
+	--------------------------------------------------------------------------------
+	-- Line Ax, Az - Bx, Bz. 
+	-- Unit Position Px, Pz
+	-- Returns closest point and progress along line `t`
+	local function GetClosestPointOnLine(Ax, Az, Bx, Bz, Px, Pz)
```

Comment:
is there no `math.foo` function for this? perhaps should be moved to commons if not